### PR TITLE
Provide a utility that parses gRPC-Web trailers from a response content

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

The trailers of a gRPC-Web response must be the last message of the response, if a response has a body.
Users don't get `tailers` via `RetryRule{WithContent}Builder.onResponseTrailers(...)`.
This could be a problem when they need `grpc-status` for retrying a gRPC-Web request.
Slack thread: https://line-armeria.slack.com/archives/C1NGPBUH2/p1592985189001300?thread_ts=1591852559.452400&cid=C1NGPBUH2

Modifications:

- Add a utility method that parses a trailers from the response content.

Result:

You can now parse a response trailers from a response body for gRPC-Web using `GrpcWebUtil.parseTrailers(response)`

Note:

I'm not sure this is the best way.
On second thought, the problems of trailers in RetryRule could be solved by providing gRPC specific RetryRule such as `GrpcRetryRule`.

### CODE REVIEW

**Comment by anuraaga** [comment]
IMO the best way to solve issues like these is to provide utilities as gRPC interceptors where they are missing in the ecosystem. As you know I think mixing HTTP and RPC is too hacky ;)

**Comment by anuraaga** [comment]
That doesn't mean logic can't be shared, but it should be by building primitives and using them natively in each layer I think, rather than shoehorning an RPC layer into a different one. If retry is too tied to HTTP right now, either it should focus on HTTP, or it could be refactored such that a common primitive is used at both HTTP decorator or gRPC interceptor, etc.

One example of creating primitives and building on it seems to be this throttling https://github.com/Netflix/concurrency-limits

**Comment by ikhoon** [comment]
> If retry is too tied to HTTP right now, either it should focus on HTTP, or it could be refactored such that a common primitive is used at both HTTP decorator or gRPC interceptor, etc.

Agreed. gRPC interceptor seems a better approach to access the headers and trailers using `Metadata`.

I think we can add a new `GrpcRetryRuleBulder`

```
class GrpcRetryRuleBuilder extends RetryRuleBuilder {

    public GrpcRetryRuleBuilder onGrpcStatus(Predicate<? super Status> statusFilter) {
        ...
    }
    public GrpcRetryRuleBuilder onMetaDataHeaders(Predicate<? super Metadata> headersFilter) {
        ...
    }
    public GrpcRetryRuleBuilder onMetaDataTrailers(Predicate<? super Metadata> trailersFilter) {
        ...
    }
}

// Interpret the given RetryRule in a RPC layer
GrpcRetryIntercepter.of(GrpcRetryRule.builder()
                                     .onMetaDataTrailers(trailers -> ...)
                                     // Less efficient than HTTP RetryRule.
                                     .onStatus(httpStatus -> ...)
                                     .onGrpcStatus(grpcStatus -> ...)
                                     .thenBackoff(myBackoff);
```

**Comment by ikhoon** [comment]
By the way, we need to refactor `RetryingClient` to use `maxTotalAttempts`, `responseTimeoutMillisForEachAttempt` and so on. Maybe it seems overkill to fix it before 1.0. 🤔

**Comment by trustin** [comment]
Yeah, I think it's a little bit late to make it available in 1.0. Probably in 1.1+?

This pull request looks good to me by itself though. What do you think, @anuraaga and @minwoox ?

**Comment by anuraaga** [comment]
Main issue with adding hacks like this are it reduces the motivation to add a better solution sooner. But if it seems important enough it's ok.

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+     *                        HttpHeaders trailers = GrpcWebUtil.parseTrailers(aggregated.content());
+     *                        // Retry if the 'grpc-status' is not equal to 0.
+     *                        return trailers != null && trailers.getInt(GrpcHeaderNames.GRPC_STATUS) != 0;
+     *                    });
+     *                })))
+     *        .build(MyGrpcStub.class);
+     * }</pre>
+     */
+    @Nullable
+    static HttpHeaders parseTrailers(HttpData response) {
```
Comment (minwoox) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java:70]: Isn't this public?

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        final ByteBuf buf;
```
Comment (minwoox) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java:71]: rnn for response?

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                    // Malformed header
+                    break;
+                }
+
+                final int length = buf.readInt();
+                if (type >> 7 == 1) {
+                    trailers = InternalGrpcWebUtil.parseGrpcWebTrailers(buf);
+                    break;
+                } else {
+                    buf.skipBytes(length);
```
Comment (minwoox) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java:93]: Could you leave a comment about what this is for?

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java (hunk 1 above)
Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcWebUtil.java:70]: Oops!🤣

**Comment by trustin** [comment]
> Main issue with adding hacks like this are it reduces the motivation to add a better solution sooner. But if it seems important enough it's ok.

Agreed. How about adding `@UnstableApi` and adding a link to a new issue dedicated to this issue, @ikhoon ?

**Comment by ikhoon** [comment]
> How about adding @UnstableApi and adding a link to a new issue dedicated to this issue, @ikhoon ?

Yap! That’s good. Let me file an issue tomorrow. :-)

**Comment by trustin** [comment]
Thanks, @ikhoon 😄

**Pull request merged**